File: Ejercicios/memento.py


class MementoEmpleado:
    def __init__(self, nombre, puesto, salario):
        self.nombre = nombre
        self.puesto = puesto
        self.salario = salario

class Empleado:
    def __init__(self):
        self.nombre = ""
        self.puesto = ""
        self.salario = 0.0

    def set_datos(self, nombre, puesto, salario):
        self.nombre = nombre
        self.puesto = puesto
        self.salario = salario

    def show(self):
        print(f"Empleado: {self.nombre}, Puesto: {self.puesto}, Salario: ${self.salario:.2f}")

    def save(self):
        return MementoEmpleado(self.nombre, self.puesto, self.salario)

    def restore(self, memento):
        self.nombre = memento.nombre
        self.puesto = memento.puesto
        self.salario = memento.salario

class History:
    def __init__(self):
        self.mementos = []

    def add(self, memento):
        self.mementos.append(memento)

    def undo(self):
        if self.mementos:
            return self.mementos.pop()
        return None


empleado = Empleado()
history = History()


empleado.set_datos("Gabriel Proaño", "Desarrollador Junior", 1200)
history.add(empleado.save())
empleado.show()


empleado.set_datos("Gabriel Proaño", "Desarrollador Senior", 2500)
empleado.show()


empleado.restore(history.undo())
empleado.show()

File: Ejercicios/Memento.java
package Ejercicios;

import java.util.Stack;


class MementoEmpleado {
    private String nombre;
    private String puesto;
    private double salario;

    public MementoEmpleado(String nombre, String puesto, double salario) {
        this.nombre = nombre;
        this.puesto = puesto;
        this.salario = salario;
    }

    public String getNombre() { return nombre; }
    public String getPuesto() { return puesto; }
    public double getSalario() { return salario; }
}


class Empleado {
    private String nombre;
    private String puesto;
    private double salario;

    public void setDatos(String nombre, String puesto, double salario) {
        this.nombre = nombre;
        this.puesto = puesto;
        this.salario = salario;
    }

    public void show() {
        System.out.println("Empleado: " + nombre + ", Puesto: " + puesto + ", Salario: $" + salario);
    }

    public MementoEmpleado save() {
        return new MementoEmpleado(nombre, puesto, salario);
    }

    public void restore(MementoEmpleado memento) {
        if (memento != null) {
            this.nombre = memento.getNombre();
            this.puesto = memento.getPuesto();
            this.salario = memento.getSalario();
        }
    }
}


class History {
    private Stack<MementoEmpleado> mementos = new Stack<>();

    public void add(MementoEmpleado memento) {
        mementos.push(memento);
    }

    public MementoEmpleado undo() {
        if (!mementos.isEmpty()) {
            return mementos.pop();
        }
        return null;
    }
}


public class Memento {
    public static void main(String[] args) {
        Empleado empleado = new Empleado();
        History history = new History();

        
        empleado.setDatos("Gabriel Proaño", "Desarrollador Junior", 1200);
        history.add(empleado.save());
        empleado.show();

        
        empleado.setDatos("Gabriel Proaño", "Desarrollador Senior", 2500);
        empleado.show();

        
        empleado.restore(history.undo());
        empleado.show();
    }
}
